src/plugins/advanced_settings/server/saved_objects/user_ui_settings_client_wrapper.ts

```diff
@@ -0,0 +1,150 @@
+
+  private isUserLevelSetting(id: string | undefined): boolean {
+    return id ? id.startsWith(CURRENT_USER_PLACEHOLDER) : false;
+  }
+
+  private normalizeDocId(id: string, request: OpenSearchDashboardsRequest, core?: CoreStart) {
+    const userName = extractUserName(request, core);
+    if (this.isUserLevelSetting(id)) {
+      if (userName) {
+        // return id.replace(CURRENT_USER, userName); // uncomment this to support version for user personal setting
+        return userName;
+      } else {
+        // security is not enabled, using global setting id
+        return id.replace(`${CURRENT_USER_PLACEHOLDER}_`, '');
+      }
+    }
+    return id;
+  }
+
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const getUiSettings = async <T = unknown>(
+      type: string,
+      id: string,
+      options: SavedObjectsBaseOptions = {}
+    ): Promise<SavedObject<T>> => {
+      if (type === 'config') {
+        const docId = this.normalizeDocId(id, wrapperOptions.request, this.core);
+        this.logger.debug(`Getting config with original: ${id} normalizeDocId: ${docId}`);
+        // user level
+        return wrapperOptions.client.get(type, docId, options);
+      }
+
+      return wrapperOptions.client.get(type, id, options);
+    };
+
+    const updateUiSettings = async <T = unknown>(
+      type: string,
+      id: string,
+      attributes: Partial<T>,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<SavedObjectsUpdateResponse<T>> => {
+      if (type === 'config') {
+        const docId = this.normalizeDocId(id, wrapperOptions.request, this.core);
+        this.logger.debug(`Getting config with original: ${id} normalizeDocId: ${docId}`);
+        // update user level settings
+        return await wrapperOptions.client.update(type, docId, attributes, options);
+      }
+      return wrapperOptions.client.update(type, id, attributes, options);
+    };
+
+    const createUiSettings = async <T = unknown>(
+      type: string,
+      attributes: T,
+      options?: SavedObjectsCreateOptions
+    ) => {
+      const userName = extractUserName(wrapperOptions.request, this.core);
+      const { id } = options || {};
+      const userLevel = this.isUserLevelSetting(id);
+
+      if (type === 'config' && id) {
+        const docId = this.normalizeDocId(id, wrapperOptions.request, this.core);
+
+        if (userLevel && userName) {
+          const permissions = {
+            permissions: new ACL()
+              .addPermission(['write'], {
+                users: [userName],
+              })
+              .getPermissions()!,
+          };
+
+          // remove buildNum from attributes
+          const { buildNum, ...others } = attributes as any;
+
+          // create with reference, the reference field will used for filter settings by user
+          return await wrapperOptions.client.create(type, others, {
+            ...options,
+            id: docId,
+            references: [
```

Comment:
Do we still need to add this `references`

Reply:
It's used to differentiate it's user setting not application settings. It's not mandatory, remove it now.